File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        self.maxWidth = 0

        def dfs(root):

            # handle leaf
            if root is None:
                return 0

            # traverse
            leftMost = dfs(root.left)
            rightMost = dfs(root.right)

            # add left + right and check if bigger than max width
            width = leftMost+rightMost
            self.maxWidth = max(self.maxWidth, width)

            return 1 + max(leftMost, rightMost)

        # main program
        dfs(root)
        return self.maxWidth

File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode InvertTree(TreeNode root) {
        
        // base case
        if(root == null) return null;

        // swap
        var tmp = root.left;
        root.left = root.right;
        root.right = tmp;

        // traverse the tree
        InvertTree(root.left);
        InvertTree(root.right);

        return root;
    }
}
